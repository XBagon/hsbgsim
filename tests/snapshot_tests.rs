use hsbgsim::*;
use insta::assert_yaml_snapshot;
use rand::SeedableRng;
use rand_xoshiro::Xoshiro256PlusPlus;
use std::collections::{BTreeMap, BTreeSet};
use test_case::test_matrix;

//Has to be the same range as in the coverage test
#[test_matrix(
    0..300
)]
pub fn random_board_snapshot(seed: u64) {
    let mut game = Game::with_seed(seed);

    let mut rng = Xoshiro256PlusPlus::seed_from_u64(!seed);

    for _ in 0..7 {
        let minion = game.instantiate_minion(MinionVariant::random(&mut rng));
        game.position_minion(minion, PlayerId::Bottom).unwrap();

        let minion = game.instantiate_minion(MinionVariant::random(&mut rng));
        game.position_minion(minion, PlayerId::Top).unwrap();
    }

    game.initialize();

    let recording = game.run_and_record_events();
    let minions: BTreeMap<_, _> =
        game.minion_instances.iter().map(|(mi_id, minion)| (mi_id, minion.variant)).collect();

    //TODO: better replay format
    assert_yaml_snapshot!(format!("random_board_snapshot_{}", seed), (minions, recording));
}

#[test]
fn coverage() {
    let mut variants_left = BTreeSet::from_iter(MinionVariant::ALL.iter());
    let variant_count = variants_left.len();

    //Has to be the same range as on the test_matrix
    for seed in 0..300 {
        let mut game = Game::with_seed(seed);

        let mut rng = Xoshiro256PlusPlus::seed_from_u64(!seed);

        for _ in 0..7 {
            let minion = game.instantiate_minion(MinionVariant::random(&mut rng));
            game.position_minion(minion, PlayerId::Bottom).unwrap();

            let minion = game.instantiate_minion(MinionVariant::random(&mut rng));
            game.position_minion(minion, PlayerId::Top).unwrap();
        }

        for (_mi_id, minion) in game.minion_instances {
            variants_left.remove(&minion.variant);
        }
    }

    assert!(variants_left.len() == 0, "{}/{} not covered.", variants_left.len(), variant_count)
}

//TODO: test implemented minions more targeted to reduce the unnecessary hight amount of tests in this state
