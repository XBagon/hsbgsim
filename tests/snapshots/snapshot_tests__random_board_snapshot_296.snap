---
source: tests/snapshot_tests.rs
expression: "(minions, recording)"
---
- ? idx: 1
    version: 1
  : Golden Rylak Metalhead
  ? idx: 2
    version: 1
  : Crackling Cyclone
  ? idx: 3
    version: 1
  : Papa Bear
  ? idx: 4
    version: 1
  : Baby Krush
  ? idx: 5
    version: 1
  : The Boommobile
  ? idx: 6
    version: 1
  : "Golden Agamaggan, the Great Boar"
  ? idx: 7
    version: 1
  : Golden King Bagurgle
  ? idx: 8
    version: 1
  : Golden Bongo Bopper
  ? idx: 9
    version: 1
  : Golden Stone Elemental
  ? idx: 10
    version: 1
  : Menagerie Mug
  ? idx: 11
    version: 1
  : Rockpool Hunter
  ? idx: 12
    version: 1
  : Impulsive Trickster
  ? idx: 13
    version: 1
  : "Ozumat's Tentacle"
  ? idx: 14
    version: 1
  : Choral Mrrrglr
- - AuraUpdate: ~
  - DeathCheck: ~
  - ProposeAttack:
      attacker:
        idx: 1
        version: 1
      defender:
        idx: 12
        version: 1
      is_outer_phase: true
  - AuraUpdate: ~
  - DeathCheck: ~
  - Attack:
      attacker:
        idx: 1
        version: 1
      defender:
        idx: 12
        version: 1
      is_outer_phase: true
  - TakeDamage:
      target:
        idx: 12
        version: 1
      amount: 6
      source:
        idx: 1
        version: 1
  - TakeDamage:
      target:
        idx: 1
        version: 1
      amount: 2
      source:
        idx: 12
        version: 1
  - AuraUpdate: ~
  - DeathCheck: ~
  - Death:
      minion:
        idx: 12
        version: 1
      source:
        idx: 4294967295
        version: 1
  - Remove:
      minion:
        idx: 12
        version: 1
  - AfterAttack:
      attacker:
        idx: 1
        version: 1
      defender:
        idx: 12
        version: 1
      is_outer_phase: true
  - AuraUpdate: ~
  - DeathCheck: ~
  - ProposeAttack:
      attacker:
        idx: 2
        version: 1
      defender:
        idx: 1
        version: 1
      is_outer_phase: true
  - AuraUpdate: ~
  - DeathCheck: ~
  - Attack:
      attacker:
        idx: 2
        version: 1
      defender:
        idx: 1
        version: 1
      is_outer_phase: true
  - TakeDamage:
      target:
        idx: 1
        version: 1
      amount: 4
      source:
        idx: 2
        version: 1
  - AuraUpdate: ~
  - DeathCheck: ~
  - AfterAttack:
      attacker:
        idx: 2
        version: 1
      defender:
        idx: 1
        version: 1
      is_outer_phase: true
  - AuraUpdate: ~
  - DeathCheck: ~
  - ProposeAttack:
      attacker:
        idx: 2
        version: 1
      defender:
        idx: 13
        version: 1
      is_outer_phase: true
  - AuraUpdate: ~
  - DeathCheck: ~
  - Attack:
      attacker:
        idx: 2
        version: 1
      defender:
        idx: 13
        version: 1
      is_outer_phase: true
  - TakeDamage:
      target:
        idx: 13
        version: 1
      amount: 4
      source:
        idx: 2
        version: 1
  - TakeDamage:
      target:
        idx: 2
        version: 1
      amount: 2
      source:
        idx: 13
        version: 1
  - AuraUpdate: ~
  - DeathCheck: ~
  - Death:
      minion:
        idx: 2
        version: 1
      source:
        idx: 4294967295
        version: 1
  - Remove:
      minion:
        idx: 2
        version: 1
  - Death:
      minion:
        idx: 13
        version: 1
      source:
        idx: 4294967295
        version: 1
  - Remove:
      minion:
        idx: 13
        version: 1
  - AfterAttack:
      attacker:
        idx: 2
        version: 1
      defender:
        idx: 13
        version: 1
      is_outer_phase: true
  - AuraUpdate: ~
  - DeathCheck: ~
  - ProposeAttack:
      attacker:
        idx: 3
        version: 1
      defender:
        idx: 8
        version: 1
      is_outer_phase: true
  - AuraUpdate: ~
  - DeathCheck: ~
  - Attack:
      attacker:
        idx: 3
        version: 1
      defender:
        idx: 8
        version: 1
      is_outer_phase: true
  - TakeDamage:
      target:
        idx: 8
        version: 1
      amount: 12
      source:
        idx: 3
        version: 1
  - TakeDamage:
      target:
        idx: 3
        version: 1
      amount: 8
      source:
        idx: 8
        version: 1
  - AuraUpdate: ~
  - DeathCheck: ~
  - Death:
      minion:
        idx: 8
        version: 1
      source:
        idx: 4294967295
        version: 1
  - Remove:
      minion:
        idx: 8
        version: 1
  - AfterAttack:
      attacker:
        idx: 3
        version: 1
      defender:
        idx: 8
        version: 1
      is_outer_phase: true
  - AuraUpdate: ~
  - DeathCheck: ~
  - ProposeAttack:
      attacker:
        idx: 4
        version: 1
      defender:
        idx: 5
        version: 1
      is_outer_phase: true
  - AuraUpdate: ~
  - DeathCheck: ~
  - Attack:
      attacker:
        idx: 4
        version: 1
      defender:
        idx: 5
        version: 1
      is_outer_phase: true
  - TakeDamage:
      target:
        idx: 4
        version: 1
      amount: 12
      source:
        idx: 5
        version: 1
  - AuraUpdate: ~
  - DeathCheck: ~
  - Death:
      minion:
        idx: 4
        version: 1
      source:
        idx: 4294967295
        version: 1
  - Remove:
      minion:
        idx: 4
        version: 1
  - AfterAttack:
      attacker:
        idx: 4
        version: 1
      defender:
        idx: 5
        version: 1
      is_outer_phase: true
  - AuraUpdate: ~
  - DeathCheck: ~
  - ProposeAttack:
      attacker:
        idx: 5
        version: 1
      defender:
        idx: 6
        version: 1
      is_outer_phase: true
  - AuraUpdate: ~
  - DeathCheck: ~
  - Attack:
      attacker:
        idx: 5
        version: 1
      defender:
        idx: 6
        version: 1
      is_outer_phase: true
  - TakeDamage:
      target:
        idx: 6
        version: 1
      amount: 12
      source:
        idx: 5
        version: 1
  - TakeDamage:
      target:
        idx: 5
        version: 1
      amount: 12
      source:
        idx: 6
        version: 1
  - AuraUpdate: ~
  - DeathCheck: ~
  - Death:
      minion:
        idx: 6
        version: 1
      source:
        idx: 4294967295
        version: 1
  - Remove:
      minion:
        idx: 6
        version: 1
  - Death:
      minion:
        idx: 5
        version: 1
      source:
        idx: 4294967295
        version: 1
  - Remove:
      minion:
        idx: 5
        version: 1
  - AfterAttack:
      attacker:
        idx: 5
        version: 1
      defender:
        idx: 6
        version: 1
      is_outer_phase: true
  - AuraUpdate: ~
  - DeathCheck: ~
  - ProposeAttack:
      attacker:
        idx: 5
        version: 1
      defender:
        idx: 10
        version: 1
      is_outer_phase: true
  - AuraUpdate: ~
  - DeathCheck: ~
  - Attack:
      attacker:
        idx: 5
        version: 1
      defender:
        idx: 10
        version: 1
      is_outer_phase: true
  - TakeDamage:
      target:
        idx: 10
        version: 1
      amount: 12
      source:
        idx: 5
        version: 1
  - TakeDamage:
      target:
        idx: 5
        version: 1
      amount: 2
      source:
        idx: 10
        version: 1
  - AuraUpdate: ~
  - DeathCheck: ~
  - Death:
      minion:
        idx: 10
        version: 1
      source:
        idx: 4294967295
        version: 1
  - Remove:
      minion:
        idx: 10
        version: 1
  - AfterAttack:
      attacker:
        idx: 5
        version: 1
      defender:
        idx: 10
        version: 1
      is_outer_phase: true
  - AuraUpdate: ~
  - DeathCheck: ~
  - ProposeAttack:
      attacker:
        idx: 14
        version: 1
      defender:
        idx: 1
        version: 1
      is_outer_phase: true
  - AuraUpdate: ~
  - DeathCheck: ~
  - Attack:
      attacker:
        idx: 14
        version: 1
      defender:
        idx: 1
        version: 1
      is_outer_phase: true
  - TakeDamage:
      target:
        idx: 1
        version: 1
      amount: 6
      source:
        idx: 14
        version: 1
  - TakeDamage:
      target:
        idx: 14
        version: 1
      amount: 6
      source:
        idx: 1
        version: 1
  - AuraUpdate: ~
  - DeathCheck: ~
  - Death:
      minion:
        idx: 14
        version: 1
      source:
        idx: 4294967295
        version: 1
  - Remove:
      minion:
        idx: 14
        version: 1
  - Death:
      minion:
        idx: 1
        version: 1
      source:
        idx: 4294967295
        version: 1
  - Remove:
      minion:
        idx: 1
        version: 1
  - AfterAttack:
      attacker:
        idx: 14
        version: 1
      defender:
        idx: 1
        version: 1
      is_outer_phase: true
  - AuraUpdate: ~
  - DeathCheck: ~
  - End: BottomWin

